txmongo/sessions.py

```diff
@@ -0,0 +1,430 @@
+from __future__ import annotations
+
+import time
+import uuid
+from dataclasses import dataclass
+from enum import Enum
+from types import TracebackType
+from typing import TYPE_CHECKING, AsyncContextManager, Optional, Type
+
+import bson
+from bson import Int64, UuidRepresentation
+from bson.raw_bson import DEFAULT_RAW_BSON_OPTIONS, RawBSONDocument
+from pymongo import WriteConcern
+from pymongo.errors import (
+    BulkWriteError,
+    ConfigurationError,
+    ConnectionFailure,
+    InvalidOperation,
+    NotPrimaryError,
+    OperationFailure,
+    PyMongoError,
+)
+
+from txmongo.pymongo_errors import _UNKNOWN_COMMIT_ERROR_CODES
+from txmongo.pymongo_internals import _reraise_with_unknown_commit
+
+if TYPE_CHECKING:
+    from txmongo.connection import ConnectionPool
+    from txmongo.types import Document
+
+
+__all__ = [
+    "ClientSession",
+]
+
+_codec_options = DEFAULT_RAW_BSON_OPTIONS.with_options(
+    uuid_representation=UuidRepresentation.STANDARD
+)
+
+
+class TxnState(Enum):
+    NONE = 1
+    STARTING = 2
+    IN_PROGRESS = 3
+    COMMITTED = 4
+    COMMITTED_EMPTY = 5
+    ABORTED = 6
+
+
+@dataclass
+class ServerSession:
+    session_id: RawBSONDocument
+    last_use: float
+
+    _transaction_id: int = 0
+    _is_dirty: bool = False
+
+    @classmethod
+    def create_with_local_id(cls) -> ServerSession:
+        return cls(
+            RawBSONDocument(
+                bson.encode({"id": uuid.uuid4()}, codec_options=_codec_options),
+                codec_options=_codec_options,
+            ),
+            time.monotonic(),
+        )
+
+    def is_about_to_expire(self, timeout_minutes: float) -> bool:
+        return time.monotonic() - self.last_use > (timeout_minutes - 1) * 60
+
+    def update_last_use(self) -> None:
+        self.last_use = time.monotonic()
+
+    @property
+    def is_dirty(self) -> bool:
+        return self._is_dirty
+
+    def mark_dirty(self) -> None:
+        self._is_dirty = True
+
+    @property
+    def transaction_id(self) -> int:
+        return Int64(self._transaction_id)
+
+    def inc_transaction_id(self) -> None:
+        self._transaction_id += 1
+
+
+@dataclass(frozen=True)
+class SessionOptions:
+    """Placeholder class for session options. Currently not used."""
+
+    pass
```

Comment:
Should we create an issue for this in github and link it here?  (Link to it from here).

Reply:
Added TODO with a link to https://github.com/twisted/txmongo/issues/303

Other session options like `snapshot` or `causalConsistency` are not required to consider Client Sessions support complete. But since we claim Transactions support, we should add at least `defaultTransactionOptions`